helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java

```diff
@@ -358,7 +358,18 @@ public void invoke(NotificationContext changeContext) throws Exception {
           configChangeListener.onConfigChange(configs, changeContext);
         } else if (_listener instanceof InstanceConfigChangeListener) {
           InstanceConfigChangeListener listener = (InstanceConfigChangeListener) _listener;
-          List<InstanceConfig> configs = preFetch(_propertyKey);
+          List<InstanceConfig> configs = Collections.emptyList();
+          if (_propertyKey.getParams().length == 0) {
+            configs = preFetch(_propertyKey);
+          } else {
+            // If there are params, that means the property key is for a specific instance
+            // and will not have children.
+            if (_preFetchEnabled) {
+              InstanceConfig config = _accessor.getProperty(_propertyKey);
+              configs =
+                  config != null ? Collections.singletonList(config) : Collections.emptyList();
+            }
+          }
```

Comment:
Logic can be simplified as:

if (_propertyKey.getParams().length > 0 && _preFetchEnabled) {
xxxx
} else {
   configs = preFetch(_propertyKey);
}

it can avoid two layer of if-else blocks.